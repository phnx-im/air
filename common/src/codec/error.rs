// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use thiserror::Error;

use super::AirCodec;

#[derive(Debug, Error)]
pub enum Error {
    #[error("Empty input slice")]
    EmptyyInputSlice,
    #[error("Invalid codec version")]
    UnknownCodecVersion,
    #[error("Codec error: {0}")]
    CodecError(#[from] CodecError),
}

#[derive(Debug, Error)]
pub struct CodecError {
    pub(super) codec_version: AirCodec,
    pub(super) error: Box<dyn std::error::Error + Send + Sync + 'static>,
}

impl std::fmt::Display for CodecError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}: {}", self.codec_version, self.error)
    }
}
