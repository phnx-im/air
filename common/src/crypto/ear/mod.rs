// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! This module and its submodules contain structs and types to facilitate (EAR)
//! encryption of other structs on the backend. See the individual submodules
//! for details.

pub mod keys;
mod trait_impls;
mod traits;

use std::marker::PhantomData;

use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};
pub use traits::{EarDecryptable, EarEncryptable, EarKey};

use aes_gcm::Aes256Gcm;
use serde::{Deserialize, Serialize};

/// This type determines the AEAD scheme used for encryption at rest (EAR) by
/// the backend.
/// TODO: Replace with a key-committing scheme.
pub type Aead = Aes256Gcm;
/// Key size of the [`Aead`] scheme
pub const AEAD_KEY_SIZE: usize = 32;
const AEAD_NONCE_SIZE: usize = 12;

#[derive(
    Clone,
    Debug,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    sqlx::Type,
)]
#[sqlx(type_name = "aead_ciphertext")]
pub struct AeadCiphertext {
    ciphertext: Vec<u8>,
    nonce: [u8; AEAD_NONCE_SIZE],
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Ciphertext<CT> {
    ct: AeadCiphertext,
    pd: PhantomData<CT>,
}

impl<CT> From<AeadCiphertext> for Ciphertext<CT> {
    fn from(aead_ciphertext: AeadCiphertext) -> Self {
        Self {
            ct: aead_ciphertext,
            pd: PhantomData,
        }
    }
}

impl<CT> From<Ciphertext<CT>> for AeadCiphertext {
    fn from(ciphertext: Ciphertext<CT>) -> Self {
        ciphertext.ct
    }
}

impl AeadCiphertext {
    pub fn new(ciphertext: Vec<u8>, nonce: [u8; AEAD_NONCE_SIZE]) -> Self {
        Self { ciphertext, nonce }
    }

    pub fn into_parts(self) -> (Vec<u8>, [u8; AEAD_NONCE_SIZE]) {
        let Self { ciphertext, nonce } = self;
        (ciphertext, nonce)
    }
}

#[cfg(any(feature = "test_utils", test))]
impl<CT> Ciphertext<CT> {
    pub fn dummy() -> Self {
        AeadCiphertext::dummy().into()
    }

    pub fn random() -> Self {
        AeadCiphertext::random().into()
    }

    pub fn flip_bit(&mut self) {
        self.ct.flip_bit();
    }
}

#[cfg(any(feature = "test_utils", test))]
impl AeadCiphertext {
    pub fn dummy() -> Self {
        Self {
            ciphertext: vec![1u8; 32],
            nonce: [1u8; AEAD_NONCE_SIZE],
        }
    }

    pub fn random() -> Self {
        use rand::Rng;

        let mut rng = rand::thread_rng();
        Self {
            ciphertext: rng.r#gen::<[u8; 32]>().into(),
            nonce: rng.r#gen::<[u8; AEAD_NONCE_SIZE]>(),
        }
    }

    fn flip_bit(&mut self) {
        let byte = self.ciphertext.pop().unwrap();
        self.ciphertext.push(byte ^ 1);
    }
}
