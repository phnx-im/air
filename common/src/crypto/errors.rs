// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use super::*;

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum KeyGenerationError {
    /// Error generating signature keypair
    #[error("Error generating signature keypair")]
    KeypairGeneration,
}

#[derive(Debug, Error)]
pub enum RandomnessError {
    #[error("Insufficient randomness")]
    InsufficientRandomness,
}

#[derive(Error, Debug, Clone)]
pub enum DecryptionError {
    /// Error decrypting ciphertext.
    #[error("Error decrypting ciphertext.")]
    DecryptionError,
    /// Error deserializing payload.
    #[error("Error deserializing payload.")]
    DeserializationError,
    /// Error serializing AAD.
    #[error("Error serializing AAD.")]
    SerializationError,
    /// Error ratcheting forward
    #[error("Error ratcheting forward")]
    RatchetForwardError(#[from] LibraryError),
}

#[derive(Error, Debug, Clone)]
pub enum EncryptionError {
    /// Not enough randomness to generate Nonce
    #[error("Not enough randomness to generate Nonce")]
    RandomnessError,
    /// Encryption error
    #[error("Encryption error")]
    EncryptionError,
    /// Codec error
    #[error("Codec error")]
    SerializationError,
    /// Error ratcheting forward
    #[error("Error ratcheting forward")]
    RatchetForwardError(#[from] LibraryError),
}
