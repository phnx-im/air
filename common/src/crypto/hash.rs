// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::marker::PhantomData;

use sha2::{Digest, Sha256};
use tls_codec::{Serialize, TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::{crypto::Labeled, identifiers::TlsStr};

pub type HashAlg = Sha256;
pub const HASH_SIZE: usize = 32;

#[derive(Debug, TlsSize, TlsSerialize)]
struct LabeledHashPayload<'a, T: Serialize> {
    operation: TlsStr<'a>,
    label: TlsStr<'a>,
    payload: T,
}

impl<T: Serialize + Labeled> LabeledHashPayload<'_, T> {
    fn new(payload: T) -> Self {
        Self {
            operation: TlsStr("Infra Hash"),
            label: TlsStr(T::LABEL),
            payload,
        }
    }
}

#[derive(TlsDeserializeBytes, TlsSerialize, TlsSize, serde::Serialize, serde::Deserialize)]
#[serde(transparent)]
pub struct Hash<T: Labeled + Sized> {
    bytes: [u8; HASH_SIZE],
    _marker: PhantomData<T>,
}

impl<T: Labeled + Sized> Hash<T> {
    #[cfg(any(feature = "test_utils", test))]
    pub fn new_for_test(mut bytes: Vec<u8>) -> Self {
        // Pad until the length is exactly HASH_SIZE
        if bytes.len() < HASH_SIZE {
            bytes.resize(HASH_SIZE, 0);
        }
        // Ensure the length is exactly HASH_SIZE
        assert_eq!(
            bytes.len(),
            HASH_SIZE,
            "Hash must be exactly {} bytes long",
            HASH_SIZE
        );
        let mut hash_bytes = [0u8; HASH_SIZE];
        hash_bytes.copy_from_slice(&bytes[..HASH_SIZE]);
        Self {
            bytes: hash_bytes,
            _marker: PhantomData,
        }
    }

    pub fn as_bytes(&self) -> &[u8; HASH_SIZE] {
        &self.bytes
    }

    pub fn into_bytes(&self) -> [u8; HASH_SIZE] {
        self.bytes
    }

    pub fn from_bytes(bytes: [u8; HASH_SIZE]) -> Self {
        Self {
            bytes,
            _marker: PhantomData,
        }
    }
}

/// Trait for types that can be hashed
///
/// Default implementation requires the type (and references to it) to implement
/// `tls_codec::Serialize` and `Labeled`.
#[allow(private_bounds)]
pub trait Hashable: HashableHelper {
    /// Returns the hash of the data as a byte array.
    fn hash(&self) -> Hash<Self> {
        HashableHelper::hash(self)
    }
}

trait HashableHelper: Labeled + Sized {
    /// Returns the hash of the data as a byte array.
    fn hash(&self) -> Hash<Self>;
}

impl<T: Labeled + Sized> HashableHelper for T
where
    for<'a> &'a T: Serialize,
{
    fn hash(&self) -> Hash<T> {
        let serialization_result = LabeledHashPayload::new(self).tls_serialize_detached();
        debug_assert!(serialization_result.is_ok(), "Serialization failed");
        let serialized_labeled_payload = serialization_result.unwrap_or_default();

        let hash_bytes = HashAlg::digest(&serialized_labeled_payload);
        Hash {
            bytes: hash_bytes.into(),
            _marker: PhantomData,
        }
    }
}

mod trait_impls {
    use sqlx::{Database, Encode, Type};

    use super::*;

    impl<T: Labeled> std::fmt::Debug for Hash<T> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "Hash<{}>({:?})", T::LABEL, self.bytes)
        }
    }

    impl<T: Labeled> PartialEq for Hash<T> {
        fn eq(&self, other: &Self) -> bool {
            self.bytes == other.bytes
        }
    }

    impl<T: Labeled> Eq for Hash<T> {}

    impl<T: Labeled> std::hash::Hash for Hash<T> {
        fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
            self.bytes.hash(state);
        }
    }

    impl<T: Labeled> Clone for Hash<T> {
        fn clone(&self) -> Self {
            Self {
                bytes: self.bytes.clone(),
                _marker: PhantomData,
            }
        }
    }

    impl<T: Labeled> Copy for Hash<T> {}

    impl<T: Labeled, DB: Database> Type<DB> for Hash<T>
    where
        Vec<u8>: Type<DB>,
    {
        fn type_info() -> <DB as Database>::TypeInfo {
            <Vec<u8> as Type<DB>>::type_info()
        }
    }

    impl<'a, T: Labeled, DB: Database> Encode<'a, DB> for Hash<T>
    where
        Vec<u8>: Encode<'a, DB>,
    {
        fn encode_by_ref(
            &self,
            buf: &mut <DB as Database>::ArgumentBuffer<'a>,
        ) -> Result<sqlx::encode::IsNull, sqlx::error::BoxDynError> {
            let bytes = self.as_bytes().to_vec();
            Encode::<DB>::encode(&bytes, buf)
        }
    }

    impl<'a, T: Labeled, DB: Database> sqlx::Decode<'a, DB> for Hash<T>
    where
        Vec<u8>: sqlx::Decode<'a, DB>,
    {
        fn decode(value: <DB as Database>::ValueRef<'a>) -> Result<Self, sqlx::error::BoxDynError> {
            let bytes_vec = <Vec<u8>>::decode(value)?;
            let bytes: [u8; HASH_SIZE] = bytes_vec
                .try_into()
                .map_err(|_| sqlx::error::BoxDynError::from("Invalid hash length"))?;
            Ok(Self {
                bytes,
                _marker: PhantomData,
            })
        }
    }
}
