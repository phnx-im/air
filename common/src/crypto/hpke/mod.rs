// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    openmls::{
        key_packages::InitKey,
        prelude::{HpkeAeadType, HpkeConfig, HpkeKdfType, HpkeKemType, HpkePublicKey},
    },
    openmls_rust_crypto::OpenMlsRustCrypto,
    openmls_traits::{
        OpenMlsProvider,
        crypto::OpenMlsCrypto,
        random::OpenMlsRand,
        types::{HpkeCiphertext, HpkePrivateKey},
    },
};
use serde::{Deserialize, Serialize};
use tls_codec::Serialize as TlsSerializeTrait;
use tracing::error;

use crate::identifiers::{ClientConfig, SealedClientReference};

use super::{
    RawKey,
    errors::{DecryptionError, EncryptionError, RandomnessError},
    secrets::SecretBytes,
};

pub mod trait_impls;

#[derive(Debug, Serialize, Deserialize)]
#[serde(transparent)]
pub struct EncryptionKey<KT> {
    key: Vec<u8>,
    _type: std::marker::PhantomData<KT>,
}

pub const HPKE_CONFIG: HpkeConfig = HpkeConfig(
    HpkeKemType::DhKem25519,
    HpkeKdfType::HkdfSha256,
    HpkeAeadType::AesGcm256,
);

impl<KT> EncryptionKey<KT> {
    fn new(bytes: Vec<u8>) -> Self {
        Self {
            key: bytes,
            _type: std::marker::PhantomData,
        }
    }

    /// Encrypt the given plaintext using this key.
    pub(crate) fn encrypt(&self, info: &[u8], aad: &[u8], plain_txt: &[u8]) -> HpkeCiphertext {
        let rust_crypto = OpenMlsRustCrypto::default();
        rust_crypto
            .crypto()
            .hpke_seal(HPKE_CONFIG, &self.key, info, aad, plain_txt)
            // TODO: get rid of unwrap
            .unwrap()
    }

    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(bytes: Vec<u8>) -> Self {
        Self::new(bytes)
    }
}

impl<KT: RawKey> EncryptionKey<KT> {
    pub fn into_bytes(self) -> Vec<u8> {
        self.key
    }

    pub fn from_bytes(bytes: Vec<u8>) -> Self {
        Self {
            key: bytes,
            _type: std::marker::PhantomData,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DecryptionKey<KT> {
    decryption_key: SecretBytes,
    #[serde(bound = "")]
    encryption_key: EncryptionKey<KT>,
}

#[cfg(any(test, feature = "test_utils"))]
impl<KT> PartialEq for DecryptionKey<KT> {
    fn eq(&self, other: &Self) -> bool {
        let decryption_key: &[u8] = self.decryption_key.as_ref();
        let other_decryption_key: &[u8] = other.decryption_key.as_ref();
        decryption_key == other_decryption_key && self.encryption_key == other.encryption_key
    }
}

#[cfg(any(test, feature = "test_utils"))]
impl<KT> Eq for DecryptionKey<KT> {}

impl<KT> DecryptionKey<KT> {
    pub fn new(decryption_key: HpkePrivateKey, encryption_key: EncryptionKey<KT>) -> Self {
        Self {
            decryption_key: decryption_key.as_ref().to_vec().into(),
            encryption_key,
        }
    }

    pub fn decrypt(
        &self,
        info: &[u8],
        aad: &[u8],
        ct: &HpkeCiphertext,
    ) -> Result<Vec<u8>, DecryptionError> {
        let rust_crypto = OpenMlsRustCrypto::default();
        rust_crypto
            .crypto()
            .hpke_open(HPKE_CONFIG, ct, &self.decryption_key, info, aad)
            .map_err(|_| DecryptionError::DecryptionError)
    }

    pub fn generate() -> Result<Self, RandomnessError> {
        let provider = OpenMlsRustCrypto::default();
        let key_seed = provider
            .rand()
            .random_array::<32>()
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        let keypair = provider
            .crypto()
            .derive_hpke_keypair(HPKE_CONFIG, &key_seed)
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        Ok(Self::new(
            keypair.private,
            EncryptionKey::new(keypair.public),
        ))
    }

    pub fn encryption_key(&self) -> &EncryptionKey<KT> {
        &self.encryption_key
    }
}

// TODO: We might want to properly fix AAD and Info at some point.
pub trait HpkeEncryptable<KT, HpkeCiphertextType: From<HpkeCiphertext>>:
    tls_codec::Serialize
{
    /// Encrypts the data with the given encryption key.
    fn encrypt(
        &self,
        encryption_key: &EncryptionKey<KT>,
        info: &[u8],
        aad: &[u8],
    ) -> HpkeCiphertextType {
        // Hiding a LibraryError behind an empty vec.
        let plain_txt = self.tls_serialize_detached().unwrap_or_default();
        encryption_key.encrypt(info, aad, &plain_txt).into()
    }
}

pub trait HpkeDecryptable<KT, HpkeCiphertextType: AsRef<HpkeCiphertext>>:
    tls_codec::DeserializeBytes + Sized
{
    fn decrypt(
        ct: HpkeCiphertextType,
        decryption_key: &DecryptionKey<KT>,
        info: &[u8],
        aad: &[u8],
    ) -> Result<Self, DecryptionError> {
        let plaintext = decryption_key.decrypt(info, aad, ct.as_ref())?;
        Self::tls_deserialize_exact_bytes(&plaintext).map_err(|e| {
            error!(%e, "Error deserializing decrypted data");
            DecryptionError::DeserializationError
        })
    }
}

pub struct JoinerInfoKeyType;
pub type JoinerInfoEncryptionKey = EncryptionKey<JoinerInfoKeyType>;

// We need this From trait, because we have to work with the hpke init key from
// the KeyPackage, which we get as HpkePublicKey from OpenMLS.
impl From<HpkePublicKey> for JoinerInfoEncryptionKey {
    fn from(value: HpkePublicKey) -> Self {
        Self::new(value.as_slice().to_vec())
    }
}

impl From<InitKey> for JoinerInfoEncryptionKey {
    fn from(value: InitKey) -> Self {
        Self::new(value.as_slice().to_vec())
    }
}

pub type JoinerInfoDecryptionKey = DecryptionKey<JoinerInfoKeyType>;

// We need this From trait, because we have to work with the hpke init key from
// the KeyPackage, which we get as HpkePrivateKey and HpkePublicKey from
// OpenMLS.
impl From<(HpkePrivateKey, InitKey)> for JoinerInfoDecryptionKey {
    fn from((sk, init_key): (HpkePrivateKey, InitKey)) -> Self {
        let vec: Vec<u8> = init_key.key().as_slice().to_vec();
        DecryptionKey::new(sk, EncryptionKey::new(vec))
    }
}

#[derive(Debug)]
pub struct ClientIdKeyType;
pub type ClientIdEncryptionKey = EncryptionKey<ClientIdKeyType>;

impl RawKey for ClientIdKeyType {}

impl ClientIdEncryptionKey {
    pub fn seal_client_config(
        &self,
        client_config: ClientConfig,
    ) -> Result<SealedClientReference, EncryptionError> {
        let bytes = client_config
            .tls_serialize_detached()
            .map_err(|_| EncryptionError::SerializationError)?;
        let ciphertext = self.encrypt(&[], &[], &bytes);
        Ok(SealedClientReference { ciphertext })
    }
}

pub type ClientIdDecryptionKey = DecryptionKey<ClientIdKeyType>;
