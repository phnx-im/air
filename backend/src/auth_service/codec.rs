// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use opaque_ke::{
    CredentialFinalization, CredentialRequest, CredentialResponse, RegistrationRequest,
    RegistrationResponse, RegistrationUpload,
};
use tls_codec::Size;

use crate::crypto::{
    OPAQUE_LOGIN_FINISH_SIZE, OPAQUE_LOGIN_REQUEST_SIZE, OPAQUE_LOGIN_RESPONSE_SIZE,
    OPAQUE_REGISTRATION_RECORD_SIZE, OPAQUE_REGISTRATION_REQUEST_SIZE,
    OPAQUE_REGISTRATION_RESPONSE_SIZE,
};

use super::{
    OpaqueLoginFinish, OpaqueLoginRequest, OpaqueLoginResponse, OpaqueRegistrationRecord,
    OpaqueRegistrationRequest, OpaqueRegistrationResponse,
};

impl Size for OpaqueRegistrationRequest {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_REGISTRATION_REQUEST_SIZE
    }
}

impl tls_codec::Serialize for OpaqueRegistrationRequest {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.client_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueRegistrationRequest {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_REQUEST_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let client_message = RegistrationRequest::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { client_message })
    }
}

impl Size for OpaqueRegistrationResponse {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_REGISTRATION_RESPONSE_SIZE
    }
}

impl tls_codec::Serialize for OpaqueRegistrationResponse {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.server_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueRegistrationResponse {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_RESPONSE_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let server_message = RegistrationResponse::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { server_message })
    }
}

impl Size for OpaqueRegistrationRecord {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_REGISTRATION_RECORD_SIZE
    }
}

impl tls_codec::Serialize for OpaqueRegistrationRecord {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.client_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueRegistrationRecord {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_RESPONSE_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let client_message = RegistrationUpload::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { client_message })
    }
}

impl Size for OpaqueLoginRequest {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_LOGIN_REQUEST_SIZE
    }
}

impl tls_codec::Serialize for OpaqueLoginRequest {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.client_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueLoginRequest {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_RESPONSE_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let client_message = CredentialRequest::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { client_message })
    }
}

impl Size for OpaqueLoginResponse {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_LOGIN_RESPONSE_SIZE
    }
}

impl tls_codec::Serialize for OpaqueLoginResponse {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.server_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueLoginResponse {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_RESPONSE_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let server_message = CredentialResponse::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { server_message })
    }
}

impl Size for OpaqueLoginFinish {
    fn tls_serialized_len(&self) -> usize {
        OPAQUE_LOGIN_FINISH_SIZE
    }
}

impl tls_codec::Serialize for OpaqueLoginFinish {
    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
        let serialized_request = self.client_message.serialize().to_vec();
        let len = writer.write(&serialized_request)?;
        Ok(len)
    }
}

impl tls_codec::Deserialize for OpaqueLoginFinish {
    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
    where
        Self: Sized,
    {
        let mut serialized_request = vec![0u8; OPAQUE_REGISTRATION_RESPONSE_SIZE];
        bytes
            .read(&mut serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        let client_message = CredentialFinalization::deserialize(&serialized_request)
            .map_err(|_| tls_codec::Error::InvalidInput)?;
        Ok(Self { client_message })
    }
}
