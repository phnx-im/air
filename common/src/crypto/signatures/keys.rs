// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::crypto::RawKey;

use super::private_keys::{SigningKey, VerifyingKey, VerifyingKeyRef};

#[derive(Debug)]
pub struct LeafVerifyingKeyType;
pub type LeafVerifyingKeyRef<'a> = VerifyingKeyRef<'a, LeafVerifyingKeyType>;

#[derive(Debug)]
pub struct QsClientVerifyingKeyType;
pub type QsClientVerifyingKey = VerifyingKey<QsClientVerifyingKeyType>;

impl RawKey for QsClientVerifyingKeyType {}

pub type QsClientSigningKey = SigningKey<QsClientVerifyingKeyType>;

#[derive(Debug)]
pub struct QsUserVerifyingKeyType;
pub type QsUserVerifyingKey = VerifyingKey<QsUserVerifyingKeyType>;

impl RawKey for QsUserVerifyingKeyType {}

pub type QsUserSigningKey = SigningKey<QsUserVerifyingKeyType>;

#[cfg(test)]
mod test {
    use crate::codec::PersistenceCodec;

    use super::*;

    #[test]
    fn qs_client_verifying_key_serde_codec() {
        let key = QsClientVerifyingKey::new_for_test(vec![1, 2, 3]);
        insta::assert_binary_snapshot!(".cbor", PersistenceCodec::to_vec(&key).unwrap());
    }

    #[test]
    fn qs_client_verifying_key_serde_json() {
        let key = QsClientVerifyingKey::new_for_test(vec![1, 2, 3]);
        insta::assert_json_snapshot!(key);
    }
}
