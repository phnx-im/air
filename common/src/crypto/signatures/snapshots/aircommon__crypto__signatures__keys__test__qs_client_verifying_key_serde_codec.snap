---
source: common/src/crypto/signatures/keys.rs
expression: "PersistenceCodec::to_vec(&key).unwrap()"
extension: cbor
snapshot_kind: binary
---
